### PR TITLE
 Implement fs copy from HTTP

### LINKED ISSUE #667: Implement fs copy from HTTP
# Description

Extend `fs copy` mode/action to get files over HTTP.

# AC

- [ ] Uses the `requests` library for data transfer.
- [ ] Uses `urllib.parse` to get protocol from URL.
- [ ] No protocol (as before) or `file://` => local filesystem.
- [ ] Attempts to use `http://` or `https://` in a destination path fail with an informative error message.
- [ ] Unit tests are added.
- [ ] Documentation is updated with one or more examples.
- [ ] Both CLI and API access is provided.

### PR DESCRIPTION
**Pull request by maddenp-noaa**
**Synopsis**

Fixes #667.

See updated docs for an example but, in short, for file copies (CLI `uw fs copy`, API `uwtools.api.fs.copy()`), interpret paths starting with `http://` or `https://` as URLs of remote files and fetch them to the local filesystem with `requests`. For example:
```
$ echo "gpl: https://www.gnu.org/licenses/gpl.txt" | uw fs copy --target-dir licenses
[2024-12-07T01:12:25]     INFO Validating config against internal schema: files-to-stage
[2024-12-07T01:12:25]     INFO 0 UW schema-validation errors found in fs config
[2024-12-07T01:12:25]     INFO File copies: Initial state: Not Ready
[2024-12-07T01:12:25]     INFO File copies: Checking requirements
[2024-12-07T01:12:25]     INFO Copy https://www.gnu.org/licenses/gpl.txt -> licenses/gpl: Initial state: Not Ready
[2024-12-07T01:12:25]     INFO Copy https://www.gnu.org/licenses/gpl.txt -> licenses/gpl: Checking requirements
[2024-12-07T01:12:26]     INFO Copy https://www.gnu.org/licenses/gpl.txt -> licenses/gpl: Requirement(s) ready
[2024-12-07T01:12:26]     INFO Copy https://www.gnu.org/licenses/gpl.txt -> licenses/gpl: Executing
[2024-12-07T01:12:26]     INFO Copy https://www.gnu.org/licenses/gpl.txt -> licenses/gpl: Final state: Ready
[2024-12-07T01:12:26]     INFO File copies: Final state: Ready
```
```
$ tree licenses
licenses
└── gpl

1 directory, 1 file
```

Unavailable upstream files are detected, leading to a no-op:

```
$ echo "gpl: https://www.gnu.org/licenses/nope.txt" | uw fs copy --target-dir licenses
[2024-12-07T01:16:41]     INFO Validating config against internal schema: files-to-stage
[2024-12-07T01:16:41]     INFO 0 UW schema-validation errors found in fs config
[2024-12-07T01:16:41]     INFO File copies: Initial state: Not Ready
[2024-12-07T01:16:41]     INFO File copies: Checking requirements
[2024-12-07T01:16:41]     INFO Copy https://www.gnu.org/licenses/nope.txt -> licenses/gpl: Initial state: Not Ready
[2024-12-07T01:16:41]     INFO Copy https://www.gnu.org/licenses/nope.txt -> licenses/gpl: Checking requirements
[2024-12-07T01:16:41]  WARNING Remote object https://www.gnu.org/licenses/nope.txt: State: Not Ready (external asset)
[2024-12-07T01:16:42]     INFO Copy https://www.gnu.org/licenses/nope.txt -> licenses/gpl: Requirement(s) not ready
[2024-12-07T01:16:42]  WARNING Copy https://www.gnu.org/licenses/nope.txt -> licenses/gpl: Final state: Not Ready
[2024-12-07T01:16:42]  WARNING File copies: Final state: Not Ready
```
```
$ tree licenses
licenses  [error opening dir]

0 directories, 0 files
```

**Type**

- [x] Documentation
- [x] Enhancement (adds new functionality)

**Impact**

- [x] This is a non-breaking change (existing functionality continues to work as expected)

**Checklist**

- [x] I have added myself and any co-authors to the PR's _Assignees_ list.
- [x] I have reviewed the documentation and have made any updates necessitated by this change.

### CODE REVIEW

File: src/uwtools/utils/tasks.py [hunk 1]
```diff
@@ -35,14 +43,26 @@ def executable(program: Union[Path, str]):
 
 
 @external
-def existing(path: Path):
+def existing(path: Union[Path, str]):
```
Comment (maddenp-noaa) [inline src/uwtools/utils/tasks.py:46]: This task function has been updated to check the existence of both local filesystem items and remote network objects. The latter will be represented as `str`s, so the function signature is updated.

For local paths, remove a `file://` prefix if present, then execute the same logic as before.

For HTTP paths, let `requests` do a `HEAD` check for existence.

For other network schemes, raise an exception.

File: src/uwtools/utils/tasks.py [hunk 2]
```diff
@@ -59,18 +79,35 @@ def file(path: Path, context: str = ""):
 
 
 @task
-def filecopy(src: Path, dst: Path):
+def filecopy(src: Union[Path, str], dst: Union[Path, str]):
```
Comment (maddenp-noaa) [inline src/uwtools/utils/tasks.py:82]: Similar to the updated `existing()` above, retain the old logic for local filesystem items, have `requests` do a `GET` on HTTP objects, and raise an exception for unsupported schemes.

**Review by christinaholtNOAA — APPROVED** [review]
LGTM.

Probably a dumb question, but this code is the back end for the `fs` tool and the `files_to_copy`/`files_to_link` sections for drivers, right?

**Comment by maddenp-noaa** [comment]
@christinaholtNOAA It's actually a great question: The core support is indeed implemented in code deeper in the library than the drivers' `files_to_copy` / `files_to_link` support. For example the common [MPAS driver code](https://github.com/maddenp-noaa/uwtools/blob/gh-667-fs-copy-http/src/uwtools/drivers/mpas_base.py#L36-L39) calls `filecopy()` from `uwtools.utils.tasks`, which now supports both local-filesystem and HTTP source paths. But the correctness checks on the source and destination paths are assumed to have already been done, and I only implemented those for the `uw fs` mode code -- not in drivers. So I think I need to generalize that. Good catch!

**Pull request merged**